Add sybil examples to more files

joaander opened the pull request:
<!-- Please confirm that your work is based on the correct branch. -->
<!-- Bug fixes should be based on *trunk-patch*. -->
<!-- Backwards compatible new features should be based on *trunk-minor*. -->
<!-- Incompatible API changes should be based on *trunk-major*. -->

## Description

<!-- Describe your changes in detail. -->
Add sybil examples to all *.py files in the `hoomd` source directory. Also reword some documentation for clarity and reduce the toctree depth.

## Motivation and context

<!--- Why is this change required? What problem does it solve? -->
Provide users with code examples that are tested and known to execute without errors.

## How has this been tested?

I executed sybil tests locally (serial CPU build).
<!--- Please describe in detail how you tested your changes. -->

<!--- Please build the sphinx documentation and check that any changes to
      documentation display properly. -->

## Change log

No additional change log is needed over that in #1574.

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

joaander committed:
Make sybil optional.

Allow pytest to succeed in environments where sybil is not available.

joaander committed:
Add code examples to Box.

joaander committed:
Show only modules in documentation module lists.

joaander committed:
Add examples to Operations.

joaander committed:
Add examples to Simulation.

joaander committed:
Add examples to Snapshot.

joaander committed:
Fix typos.

joaander committed:
Document examples in State.

joaander committed:
Add examples to Communicator.

joaander committed:
Revise Logger documentation.

joaander committed:
Add examples to mesh and trigger.

joaander committed:
Add examples to Trigger and subclasses.

joaander committed:
Add examples to variants.

joaander committed:
Add code examples to wall.py.

joaander committed:
Ask sybil to parse all .py files.

* Exclude those that we have not yet updated to work with sybil.
* Change R""" to r""" to avoid sybil parsing errors.

joaander committed:
Run pre-commit.

joaander committed:
Fix test failure with BUILD_MD=off.

joaander committed:
Fix doc build.

joaander committed:
Use GPU device to test gpu_local_snapshot.

joaander committed:
Fix grammar error.

joaander committed:
Fix syntax error.

b-butler commented on hoomd/write/gsd.py:
```diff
@@ -55,7 +55,7 @@ class GSD(Writer):
     Args:
         trigger (hoomd.trigger.trigger_like): Select the timesteps to write.
-        filename (str): File name to write.
+        filename (`str`): File name to write.
```
Given the explicit conversion to a string, we now would accept `PurePath` objects as well.

joaander replied:
Do most HOOMD users know what a `PurePath` object is? We also accept any time that will convert to a filename with `str()` - e.g. `filename=12345`. What is the universal way of documenting that we accept any type that will convert to a string?

b-butler commented on hoomd/simulation.py:
```diff
@@ -187,6 +215,12 @@ def create_state_from_gsd(self,
+
+        .. code-block:: python
+
+            simulation.create_state_from_gsd(filename = path / 'file.gsd')
```
I think that adding a variable `filename` to the invisible code block would be more understandable for the new Python user as `Path` objects are not well known from my estimation.

joaander replied:
I'm not certain that `filename=filename` would be any less confusing to novice users.

SchoeniPhlippsn replied:
I agree that `filename = filename` might not be the best choice as well. Maybe we can rename `path` to something more specific such as `path_to_input_file`.

b-butler replied:
It isn't so much the name, but the `Path("/path/to/tmp/dir") / "file.gsd")` syntax that I thought might be advanced.

b-butler commented on hoomd/md/update.py:
```diff
@@ -220,7 +220,7 @@ def _attach_hook(self):
     @log(category="scalar", requires_run=True)
     def summed_exchanged_momentum(self):
-        R"""Returned the summed up exchanged velocity of the full simulation."""
+        """Returned the summed up exchanged velocity of the full simulation."""
```
```suggestion
        """Returns the summed up exchanged velocity of the full simulation."""
```

b-butler commented on hoomd/logging.py:
```diff
@@ -657,14 +625,20 @@ class they come from. For instance, the `hoomd.md.pair.LJ` class has a
+
+    .. code-block:: python
+
+        logger = hoomd.logging.Logger()
```
See my PR #1588 where I added example logger constructions. We should choose one to keep.

joaander replied:
If you think the additional examples are helpful, we can keep them. I have been avoiding code comments in examples. If additional explanation or a header is needed, put that outside the code block. For example: https://github.com/glotzerlab/hoomd-blue/blob/4e30e11957421a349da1371a6980c12497d058cf/hoomd/md/methods/methods.py#L107-L123

joaander replied:
`logger` is needed to run the later examples in the file. We will add more examples (if any) during merge conflict resolution.

b-butler reviewed (commented):
The PR looks good overall I just have a few small questions/comments.

joaander committed:
Workaround failing gpu_local_snapshot_test.

The example simulation had 0 particles on rank 0, leading to
(0,) shape arrays instead of the expected (0,3) shape array.

joaander committed:
Fix typo.

Co-authored-by: Brandon Butler <butlerbr@umich.edu>

SchoeniPhlippsn commented on hoomd/box.py:
```diff
@@ -176,15 +160,19 @@ class Box:
     `Box` has factory methods to enable easier creation of boxes: `cube`,
-    `square`, `from_matrix`, and `from_box`. See the method documentation for
-    usage.
+    `square`, `from_matrix`, and `from_box`. See the each methods documentation
```
```suggestion
    `square`, `from_matrix`, and `from_box`. See the methods documentation
```

SchoeniPhlippsn commented on hoomd/box.py:
```diff
@@ -455,12 +546,17 @@ def to_matrix(self):
         defines the box.
 
-        .. code-block:: python
+        .. code-block::
```
In line 227 you removed `.. code-block::` completely. I would make it consistent throughout the whole file.

joaander replied:
I made this change intentionally. This block is not a code example, it is a monospace font formatted matrix. I removed the sphinx `.. code-block::` directive and used the restructured text literal block instead: https://docutils.sourceforge.io/0.4/docs/ref/rst/restructuredtext.html#literal-blocks

All visible code examples that execute in sybil must be preceded by the sphinx block `.. code-block:: python`: https://sybil.readthedocs.io/en/latest/rest.html#python - other `.. code-block::` blocks are ignored by sybil.

joaander replied:
I added a math formatted matrix.

SchoeniPhlippsn commented on hoomd/logging.py:
```diff
@@ -681,13 +655,13 @@ def categories(self):
     @property
     def string_categories(self):
-        """`list` of `str`: A list of the string names of the allowed \
+        """list[str]: A list of the string names of the allowed \
```
```suggestion
        """`list` [`str`]: A list of the string names of the allowed \
```

joaander replied:
These are parsed correctly as is. The napoleon engine in sphinx understands these type annotations and expands `list` and `str` to links. We need explicit links only in cases where we add additional text to the type e.g. "`list` [`str`], optional".

SchoeniPhlippsn commented on hoomd/state.py:
```diff
@@ -348,32 +363,74 @@ def set_snapshot(self, snapshot):
     @property
     def particle_types(self):
-        """list[str]: List of all particle types in the simulation state."""
+        """list[str]: List of all particle types in the simulation state.
```
```suggestion
        """`list` [`str`]: List of all particle types in the simulation state.
```
Same for all the following

SchoeniPhlippsn reviewed (commented):
PR looks good. Only some minor comments.

joaander committed:
Use gsd_filenme in place of platform independent path operations.

joaander committed:
Fix typo.

joaander committed:
Use math formatted matrix.

joaander committed:
Clarify filename argument types accepted.

b-butler reviewed (approved):
LGTM.
